engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java

```diff
@@ -0,0 +1,105 @@
+package org.enso.interpreter.test;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.file.Paths;
+import java.util.Map;
+import org.enso.polyglot.RuntimeOptions;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Language;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PolyglotErrorTest {
+  private Context ctx;
+  private Value panic;
+
+  public static String bar(Object o) {
+    return "[[" + o + "]]";
+  }
+
+  @Before
+  public void prepareCtx() throws Exception {
+    this.ctx = Context.newBuilder()
+      .allowExperimentalOptions(true)
+      .allowIO(true)
+      .allowAllAccess(true)
+      .logHandler(new ByteArrayOutputStream())
+      .option(
+        RuntimeOptions.LANGUAGE_HOME_OVERRIDE,
+        Paths.get("../../distribution/component").toFile().getAbsolutePath()
+      ).build();
+    final Map<String, Language> langs = ctx.getEngine().getLanguages();
+    assertNotNull("Enso found: " + langs, langs.get("enso"));
+
+    var code = """
+    import Standard.Base.Panic.Panic
+    import Standard.Base.Error.Error
+    import Standard.Base.Error.Illegal_State.Illegal_State
+    polyglot java import org.enso.interpreter.test.PolyglotErrorTest
+
+    type TypeCa
+        Ca x
+
+        to_text : Text
+        to_text self = "<<Ca "+self.x.to_text+">>"
+
+    type TypeCb
+        Cb x
+
+        to_text : Text
+        to_text self = Error.throw (Illegal_State.Error "B")
+
+    type TypeCc
+        Cc x
+
+        to_text : Text
+        to_text self = Panic.throw (Illegal_State.Error "C")
+
+    panic x = case x of
+        1 -> panic1
+        2 -> panic2
+        _ -> panic3
+
+    panic1 = PolyglotErrorTest.bar (TypeCa.Ca 'x')
```

Comment:
@radeusgd, if you want me to cover more cases, feel free to expand the test and I do my best to deal with the additional situations.

Reply:
The test cases are ok, but I'm not sure if the behavior is what we want.

Of course, if there is no way to throw a panic in case `to_text` failed with a panic/error, then so be it. But looking at the code above, can we throw instead of returning the fallback? If not, could you please explain to me why not?

I'd value @jdunkerley's input here as well. I think in some cases a fallback text representation is actually useful (for example when I'm trying to debug something, if I get an error in `to_text` I won't even immediately know where it happened without some fallback). On the other hand, from the libraries perspective I think it is very important to be able to detect these errors.

Consider case the user saves some custom object in a Table, whose `to_text` is broken and is erroring. The user is serializing such a table to a CSV file. I think I'd rather get an error/warning saying that some objects failed to serialize (and e.g. in warning mode I'd replace the problematic entries with `null`/`""` instead of some fallback which may not always be appropriate).

Ideally, I think it would be best if we could throw some special exception `Unexpected_Failure_In_To_Text` which would contain within its message/payload that mentioned fallback text representation - this way we will kill 2 birds with one stone - we'll still be able to handle the errors in our libraries, but for example when debugging we will have at least some concept of what is the object whose `to_text` was broken.

Reply:
I've just started [discussion on GraalVM slack channel](https://graalvm.slack.com/archives/CNQSB2DHD/p1673413554022339?thread_ts=1673412770.137559&cid=CNQSB2DHD) to bring this issue to GraalVM guys attention.

Reply:
Thanks for bringing up the topic on Slack.

The changes to the implementation made me realise it would be good to at least handle one more case:

```
type TypeCd
    Cd x

    to_text self = 42

panic4 = PolyglotErrorTest.bar (TypeCd.Cd Nothing)
```

I don't have a very particular requirement of what the message should be, but it should clearly indicate to the user that an unexpected type has been provided.

I'd suggest something along the lines of:
```
[Error in method `to_text` of [TypeCd.Cd Nothing]: Type_Error: expected ??? to be Text but got 42.]
```

We probably need to amend the `UnexpectedResultException` branch and change the error message depending on if `result` is a dataflow error or if it contains any other value.

To make sure the logic is correctly implemented ideally we'd also have these tests:

```
type TypeCe
    Cd x

    to_text self =
        Warning.attach "Some random warning" self.x

text5 = PolyglotErrorTest.bar (TypeCd.Ce "Foo")
panic6 = PolyglotErrorTest.bar (TypeCd.Ce 44)
```

I'd then expect `text5` to just be equal to `Foo` - i.e. I want to check that attaching a warning within the `to_text` call does not break the interop `toString`.

And `panic6` should return a result consistent with `panic4` - i.e. it should say that it expected a Text but got a `44`. It should not display something like `WithWarnings@deadbeef`.

Reply:
Tests up to `panic6` implemented in [4997b24](https://github.com/enso-org/enso/pull/4034/commits/4997b246a92e709ad0290c6ca6dadca97603a00c).

Reply:
Thanks!